File: totaleconomy-core/src/main/java/com/ericgrandt/totaleconomy/data/CurrencyData.java
package com.ericgrandt.totaleconomy.data;

import com.ericgrandt.totaleconomy.exception.CurrencyNotFoundException;
import com.ericgrandt.totaleconomy.exception.MissingDefaultCurrencyException;
import com.ericgrandt.totaleconomy.model.TECurrency;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class CurrencyData {
    public TECurrency getDefaultCurrency(Connection conn) throws SQLException {
        var query = "SELECT code, name, plural_name, symbol, fractional_digits, starting_balance, is_default FROM te_currency WHERE is_default IS TRUE LIMIT 1";
        try (PreparedStatement stmt = conn.prepareStatement(query)) {
            try (ResultSet rs = stmt.executeQuery()) {
                if (rs.next()) {
                    return new TECurrency(
                        rs.getString("code"),
                        rs.getString("name"),
                        rs.getString("plural_name"),
                        rs.getString("symbol"),
                        rs.getInt("fractional_digits"),
                        rs.getBigDecimal("starting_balance"),
                        rs.getBoolean("is_default")
                    );
                }
            }
        }

        throw new MissingDefaultCurrencyException();
    }

    // TODO: Return an Optional<TECurrency> instead of throwing the exception
    public TECurrency getCurrency(Connection conn, String currencyCode) throws SQLException {
        var query = "SELECT code, name, plural_name, symbol, fractional_digits, starting_balance, is_default FROM te_currency WHERE code = ?";
        try (PreparedStatement stmt = conn.prepareStatement(query)) {
            stmt.setString(1, currencyCode);

            try (ResultSet rs = stmt.executeQuery()) {
                if (rs.next()) {
                    return new TECurrency(
                        rs.getString("code"),
                        rs.getString("name"),
                        rs.getString("plural_name"),
                        rs.getString("symbol"),
                        rs.getInt("fractional_digits"),
                        rs.getBigDecimal("starting_balance"),
                        rs.getBoolean("is_default")
                    );
                }
            }
        }

        throw new CurrencyNotFoundException();
    }
}

File: totaleconomy-core/src/main/kotlin/com/ericgrandt/totaleconomy/data/CurrencyData.kt
package com.ericgrandt.totaleconomy.data

import com.ericgrandt.totaleconomy.data.table.CurrencyTable
import com.ericgrandt.totaleconomy.mapper.toTECurrency
import com.ericgrandt.totaleconomy.model.TECurrency
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.runCatching
import org.jetbrains.exposed.v1.core.eq
import org.jetbrains.exposed.v1.jdbc.selectAll

class CurrencyData {
    fun getDefaultCurrency(): Result<TECurrency, Throwable> {
        return runCatching {
            CurrencyTable
                .selectAll()
                .where { CurrencyTable.isDefault eq true }
                .single()
                .toTECurrency()
        }
    }

    fun getCurrencyList(): Result<List<TECurrency>, Throwable> {
        return runCatching {
            CurrencyTable.selectAll().map {
                it.toTECurrency()
            }
        }
    }
}
